ykcompile/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{
+    Constant, ConstantInt, Operand, Rvalue, Statement, TirOp, TirTrace, UnsignedInt,
+};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer,
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a `u64`.
+        let func: fn() -> u64 = unsafe { mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0))) };
+        func()
+    }
+}
+
+/// The `TraceCompiler` takes a `SIRTrace` and compiles it to machine code. Returns a `CompiledTrace`.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    /// Contains the list of currently available registers.
+    available_regs: Vec<u8>,
+    /// Maps locals to their assigned registers.
+    assigned_regs: HashMap<u32, u8>,
+}
+
+impl TraceCompiler {
+    fn local_to_reg(&mut self, l: u32) -> u8 {
+        // This is a really dumb register allocator, which runs out of available registers after 7
+        // locals. We can do better than this by using StorageLive/StorageDead from the MIR to free
+        // up registers again, and allocate additional locals on the stack. Though, ultimately we
+        // probably want to implement a proper register allocator, e.g. graph colouring or linear
+        // scan.
+        if l == 0 {
+            0
+        } else {
+            let reg = self
+                .available_regs
+                .pop()
+                .expect("Can't allocate more than 7 locals yet!");
+            *self.assigned_regs.entry(l).or_insert(reg)
+        }
+    }
+
+    /// Move constant `c` of type `usize` into local `a`.
+    pub fn mov_local_usize(&mut self, local: u32, cnst: usize) {
+        let reg = self.local_to_reg(local);
+        dynasm!(self.asm
+            ; mov Rq(reg), cnst as i32
+        );
+    }
+
+    /// Move constant `c` of type `u8` into local `a`.
+    pub fn mov_local_u8(&mut self, local: u32, cnst: u8) {
+        let reg = self.local_to_reg(local);
+        dynasm!(self.asm
+            ; mov Rq(reg), cnst as i32
+        );
+    }
+
+    /// Move local `var2` into local `var1`.
+    fn mov_local_local(&mut self, l1: u32, l2: u32) {
+        let lreg = self.local_to_reg(l1);
+        let rreg = self.local_to_reg(l2);
+        dynasm!(self.asm
+            ; mov Rq(lreg), Rq(rreg)
+        );
+    }
+
+    fn nop(&mut self) {
+        dynasm!(self.asm
+            ; nop
+        );
+    }
+
+    fn c_mov_int(&mut self, local: u32, constant: &ConstantInt) {
```

Comment:
I'm still not certain if these function names are descriptive enough.

Here we have `mov_int` to mean move a constant integer to a register.

Suppose we need to move a constant into a memory location. What would we call that?

Do we need to put the operand types in the function name? E.g.

 * `c_mov_const_to_reg`
 * `c_mov_const_to_mem`

maybe we should have a method `store_const_to_local` which calls one of the above based upon where the local currently lives (register or memory).

Thoughts?

Reply:
I wouldn't worry *too* much about this now: as we expand functionality, a better naming scheme might make itself apparent.

Reply:
OK. I'm happy to continue taking into account that these function names are not fixed.

Reply:
I agree with both of you. I think be probably should make the function names more descriptive. But I also want to get this PR merged now and think we don't need to worry about this just yet.